Arm/State.lean

```diff
@@ -662,3 +662,318 @@ theorem write_mem_irrelevant :
   simp [read_mem, write_mem, store_write_irrelevant]
 
 end Memory
+
+section NewMemory
+/-!
+# New definitions for the memory model
+
+For freedom in experimenting with definitions,
+we define our own version of `read_mem` and `write_mem`,
+called `read_mem'` and `write_mem`'.
+These operate directly on the memory, rather than the `ArmState`.
+We prove their equivalence to the existing definitions
+(`read_mem_eq_read_mem'`, `write_mem_eq_write_mem'`).
+
+Furthermore, we define equivalences to `BitVec.extractLsByte`
+to ease reasoning about byte-level manipulation.
+As an easy corollary, we get `getLsb` theorems on these to allow
+`omega` based reasoning about bit-level values of memory.
+-/
+
+abbrev Memory := Store (BitVec 64) (BitVec 8)
+
+/--
+A variant of `read_mem` that directly talks about writes to memory, instead of over the entire `ArmState`
+-/
+def read_mem' (addr : BitVec 64) (m : Memory) : BitVec 8 :=
+  read_store addr m
+
+theorem read_mem_eq_read_mem' : read_mem addr s = read_mem' addr s.mem := rfl
+
+@[simp]
+theorem getLsb_read_mem' : (read_mem' addr s).getLsb i = (s addr).getLsb i :=
+  rfl
+
+def read_mem_bytes' (n : Nat) (addr : BitVec 64) (s : Memory) : BitVec (n * 8) :=
+  match n with
+  | 0 => 0#0
+  | n' + 1 =>
+    let byte := read_mem' addr s
+    let rest := read_mem_bytes' n' (addr + 1#64) s
+    have h : n' * 8 + 8 = (n' + 1) * 8 := by simp_arith
+    BitVec.cast h (rest ++ byte)
```

Comment:
```suggestion
def read_mem_bytes' (n : Nat) (addr : BitVec 64) (m : Memory) : BitVec (n * 8) :=
  match n with
  | 0 => 0#0
  | n' + 1 =>
    let byte := read_mem' addr m
    let rest := read_mem_bytes' n' (addr + 1#64) m
    have h : n' * 8 + 8 = (n' + 1) * 8 := by simp_arith
    BitVec.cast h (rest ++ byte)
```

Let's consistently use `m` for memory and `s` for state to make it easy to immediately spot which definitions we're talking about. Please also check this for lemmas